sharding/indexHashedNodesCoordinator.go

```diff
@@ -239,37 +351,140 @@ func (ihgs *indexHashedNodesCoordinator) GetSelectedPublicKeys(selection []byte,
 	validatorsPubKeys := make(map[uint32][][]byte)
 
-	for shardId, shardEligible := range ihgs.nodesMap {
+	ihgs.mutNodesConfig.RLock()
+	nodesConfig, ok := ihgs.nodesConfig[epoch]
+	ihgs.mutNodesConfig.RUnlock()
+
+	if !ok {
+		return nil, ErrEpochNodesConfigDesNotExist
+	}
+
+	nodesConfig.mutNodesMaps.RLock()
+	defer nodesConfig.mutNodesMaps.RUnlock()
+
+	for shardId, shardEligible := range nodesConfig.eligibleMap {
 		for i := 0; i < len(shardEligible); i++ {
-			validatorsPubKeys[shardId] = append(validatorsPubKeys[shardId], ihgs.nodesMap[shardId][i].PubKey())
+			validatorsPubKeys[shardId] = append(validatorsPubKeys[shardId], nodesConfig.eligibleMap[shardId][i].PubKey())
 		}
 	}
 
-	return validatorsPubKeys
+	return validatorsPubKeys, nil
 }
 
 // GetValidatorsIndexes will return validators indexes for a block
-func (ihgs *indexHashedNodesCoordinator) GetValidatorsIndexes(publicKeys []string) []uint64 {
-	validatorsPubKeys := ihgs.GetAllValidatorsPublicKeys()
+func (ihgs *indexHashedNodesCoordinator) GetValidatorsIndexes(
+	publicKeys []string,
+	epoch uint32,
+) ([]uint64, error) {
 	signersIndexes := make([]uint64, 0)
 
+	validatorsPubKeys, err := ihgs.GetAllValidatorsPublicKeys(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	ihgs.mutNodesConfig.RLock()
+	nodesConfig := ihgs.nodesConfig[epoch]
+	ihgs.mutNodesConfig.RUnlock()
+
 	for _, pubKey := range publicKeys {
-		for index, value := range validatorsPubKeys[ihgs.shardId] {
+		for index, value := range validatorsPubKeys[nodesConfig.shardId] {
 			if bytes.Equal([]byte(pubKey), value) {
 				signersIndexes = append(signersIndexes, uint64(index))
 			}
 		}
 	}
 
-	return signersIndexes
+	if len(publicKeys) != len(signersIndexes) {
+		log.Error("len pubKeys = %d, len signersIndexes = %d", len(publicKeys), len(signersIndexes))
+		return nil, ErrNotInvalidNumberPubKeys
+	}
+
+	return signersIndexes, nil
+}
+
+// EpochStartPrepare wis called when an epoch start event is observed, but not yet confirmed/committed.
+// Some components may need to do some initialisation on this event
+func (ihgs *indexHashedNodesCoordinator) EpochStartPrepare(metaHeader data.HeaderHandler) {
+	randomness := metaHeader.GetRandSeed()
+	newEpoch := metaHeader.GetEpoch()
+
+	if newEpoch != ihgs.currentEpoch+1 {
+		return
+	}
+
+	ihgs.mutNodesConfig.RLock()
+	nodesConfig, ok := ihgs.nodesConfig[ihgs.currentEpoch]
+	ihgs.mutNodesConfig.RUnlock()
+
+	if !ok {
+		log.Error("no configured epoch found")
+		return
+	}
+
+	// TODO: update the new nodes and leaving nodes as well
+	shufflerArgs := ArgsUpdateNodes{
+		Eligible: nodesConfig.eligibleMap,
+		Waiting:  nodesConfig.waitingMap,
+		NewNodes: make([]Validator, 0),
+		Leaving:  make([]Validator, 0),
+		Rand:     randomness,
+		NbShards: nodesConfig.nbShards,
+	}
+
+	eligibleMap, waitingMap, _ := ihgs.shuffler.UpdateNodeLists(shufflerArgs)
+
+	_ = ihgs.SetNodesPerShards(eligibleMap, waitingMap, newEpoch)
```

Comment:
this just adds the config for the new epoch in the map, it doesn't remove the old one. So in case it will be committed, and there are already blocks proposed by the new validators that are confirming the epoch change, then we will have the information to validate the signature for those blocks. In short, to validate the epoch change is committed we already need the configuration of nodes for this new epoch, this is why I need to create it, just in case.